server/client.go

```diff
@@ -943,7 +960,36 @@ func (c *client) typeString() string {
 	return "Unknown Type"
 }
 
+func (c *client) closeConnectionWithEvent(reason string) {
+
+	c.mu.Lock()
+
+	if c.nc == nil {
+		c.mu.Unlock()
+		return
+	}
+
+	srv := c.srv
+	isClient := c.typ == CLIENT
+
+	c.mu.Unlock()
+
+	// do not publish on route errors.  That will be a seperate type of notification.
+	if isClient {
+		srv.mu.Lock()
```

Comment:
Looking at this, I think you would remove and simply call the publish event function. However, in `createConnectEvent`, when locking the client, return if `c.nc` is nil there, otherwise you may get a panic.